lib/python/rose/config_diff.py

```diff
@@ -0,0 +1,234 @@
+    string_file = StringIO.StringIO()
+    rose.config.dump(id_node, target=string_file)
+    string_file.seek(0)
+    return string_file.read()
+
+
+def main():
+    """Implement the "rose config-diff" command."""
+    opt_parser = rose.opt_parse.RoseOptionParser()
+    opt_parser.add_my_options("diff_tool", "graphical",
+                              "ignore", "meta_path", "properties")
+    my_sys_args = list(sys.argv)
+    if "--" in sys.argv:
+        my_sys_args = sys.argv[:sys.argv.index("--")]
+        diff_args = sys.argv[sys.argv.index("--") + 1:]
+    else:
+        diff_args = []
+
+    opts, args = opt_parser.parse_args(my_sys_args[1:])
+    rose.macro.add_site_meta_paths()
+    rose.macro.add_env_meta_paths()
+    rose.macro.add_opt_meta_paths(opts.meta_path)
+
+    if len(args) != 2:
+        sys.exit(opt_parser.get_usage())
+
+    config_loader = rose.config.ConfigLoader()
+
+    if opts.properties is None:
+        properties = _DEFAULTS.PROPERTIES.split(",")
+    else:
+        properties = opts.properties.split(",")
+
+    ignore_regexes = expand_shorthand_in_ignore_regexes(
+        opts.ignore_patterns)
+
+    output_filenames = []
+    config_type = rose.SUB_CONFIG_NAME
+    for path in args:
+        if path == "-":
+            continue
+        config_type = path.split(os.sep)[-1]
+    for path in args:
+        if path == "-":
+            path = sys.stdin
+            filename = config_type
+            directory = None
+        else:
+            filename = path.split(os.sep)[-1]
+            directory = path.rsplit(os.sep, 1)[0]
+        config = config_loader.load_with_opts(path)
+        meta_config_tree = rose.macro.load_meta_config_tree(
+            config, directory=directory, config_type=filename)
+        if meta_config_tree is None:
+            meta_config = rose.config.ConfigNode()
+        else:
+            meta_config = meta_config_tree.node
+        annotated_config = annotate_config_with_metadata(
+            config, meta_config, ignore_regexes=ignore_regexes,
+            metadata_properties=properties
+        )
+        output_dir = tempfile.mkdtemp()
+        output_path = os.path.join(output_dir, filename)
+        rose.config.dump(annotated_config, target=output_path)
+        output_filenames.append(output_path)
+    popener = rose.popen.RosePopener()
+    cmd_opts_args = diff_args + output_filenames
+    if opts.diff_tool is None:
+        if opts.graphical_mode:
+            diff_cmd = popener.get_cmd("gdiff_tool", *cmd_opts_args)
+        else:
+            diff_cmd = popener.get_cmd("diff_tool", *cmd_opts_args)
+    else:
+        diff_cmd = shlex.split(opts.diff_tool) + cmd_opts_args
+    return_code = 1
+    try:
+        return_code, stdout, stderr = popener.run(*diff_cmd)
+        sys.stdout.write(stdout)
+        sys.stderr.write(stderr)
+    finally:
+        for path in output_filenames:
+            shutil.rmtree(os.path.dirname(path))
+    sys.exit(return_code)
```

Comment:
```python
try:
    popener.run_ok(*diff_cmd, stdout=sys.stdout, stderr=sys.stderr)
except RosePopenError as exc:
    sys.exit(exc.rc)
finally:
    for path in output_filenames:
        shutil.rmtree(os.path.dirname(path))
```

(May also want to use equivalent in `rose.fs_util` instead of `shutil.rmtree`.)